Fix ID generation of foreign keys

cb8 opened the pull request:
this patch is an alternate approach to #7003 which tries to fix #7002. however, it applies to the `master` branch. so far, i didn't clean up my test cases. in case this patch will get merged despite/instead of #7003, i will push them for review.

vhenzl committed:
Add failing tests for #6531

Tests are based on examples from "Composite and Foreign Keys as Primary Key" tutorial:
http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/tutorials/composite-primary-keys.html

cb8 committed:
Fix ID generation of foreign keys

lcobucci commented:
@cb8 thanks for your patch, we just merged in the fix for 2.6 and were porting it to master. I'll check your changes to see if it works better for the new version of the code.

lcobucci commented:
@cb8 your changeset makes the tests of #6701 pass but breaks `SecondLevelCacheCompositePrimaryKeyWithAssociationsTest` because it tries to update a read only entity, I'll investigate further this week but it would be nice if you could check it as well.

cb8 committed:
Avoid flattening of post-insert IDs

When setting deferred ID values, the original data should contain
complete objects (as it would with non-deferred IDs).

cb8 commented:
thank you for the review. the deferred IDs were inadvertently flattened when they were added to the original data of the entity. as far as i could see, the failing tests on Travis are not caused by the patch (but don't hold me to that ;)).

lcobucci commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -962,9 +962,9 @@ private function executeInserts(ClassMetadata $class) : void
+                $id  = $persister->getIdentifier($entity);
 
-                $this->entityIdentifiers[$oid]  = $id;
+                $this->entityIdentifiers[$oid]  = $this->em->getIdentifierFlattener()->flattenIdentifier($class, $id);
```
@guilhermeblanco this looks ok, but I'd like your eyes here

lcobucci reviewed (changes requested):
Some small changes, I'll take care of these in a bit. @cb8 thanks a lot for your help!

lcobucci commented:
@cb8 🚢 thanks for your contribution!
